Frameworks/include/NSOrderedSetInternal.h

```diff
@@ -0,0 +1,28 @@
+//******************************************************************************
+//
+// Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+#pragma once
+
+#import <Starboard.h>
+#import <Foundation/NSOrderedSet.h>
+#import <map>
+#import <vector>
+
+@interface NSOrderedSet () {
+@protected
+    std::vector<StrongId<id>> setContainer;
+    std::map<id, std::vector<StrongId<id>>::iterator> mapContainer;
```

Comment:
A combination of an array (for indexing) and a set (for membership checks) seems fine to me. It's not optimal, because you have to keep two data structures with duplicated data, and make sure they stay in sync, but I don't object too much.

NSOrderedSet seems like an odd data structure to me. To get sub-linear membership checks, where the objects are not necessarily comparable, we're probably talking about a hash table. A hash table doesn't usually provide an ordering, but it could, if you modified it so that the entries in the hash table are all joined in a big linked list. That would give you fast membership checks and an ordered enumeration, but slow lookup of an element by index. I don't really know if that's what Apple had in mind.

---
In reply to: [62575598](https://github.com/Microsoft/WinObjC/pull/482#discussion_r62575598) [](ancestors = 62575598)